### PR TITLE
FS-860 add email page

### PR DESCRIPTION
**Pull request by SentientHousePlant**
# Summary
This is a fairly straight forward addition to the discovery frontend. The following changes have been made:

- A new "enter your email" page and a confirmation page ( The confirmation page will be replaced by @ramsharma-prog in his next pr ) have been added between picking a round and being directed to the relevant application.
- When an email is entered a new account is created in the account store (unless a 204 is returned).

### CODE REVIEW

**Comment by SentientHousePlant** [comment]
Sorry folks, I forgot some docstrings. Adding now.

File: app/discovery/models/data.py [hunk 1]
```diff
@@ -193,3 +195,72 @@ def list_data(json_data, data_func):
...
+    store.
+
+    Args:
+        email_address (str): The email address we wish
+        to create a new account with.
+
+    Returns:
+        Tuple[int, response data]: A tuple containing
+        the status code and the  response data as a str
+        - if successful.
```
Comment (ramsharma-prog) [inline app/discovery/models/data.py:250]: nice docstring

File: app/discovery/routes.py [hunk 2]
```diff
@@ -75,4 +78,58 @@ def fund_rounds(fund_id):
...
+
+        params = { "application_url" : application_url, "email" : form.email.data}
+
+        req = PreparedRequest()
+
+        root_url = request.root_url
+
+        url = root_url + url_for("discovery_bp.account_info_route")
+
+        req.prepare_url(url, params)
```
Comment (ramsharma-prog) [inline app/discovery/routes.py:104]: I like the PreparedRequest method 😄

File: app/discovery/templates/debug_continue.html [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+{% if account_exists == True%}
+<h2 class="govuk-heading-m">This account exists! Here's the data: </h2>
+<p class="govuk-body">{{account_data|string}}</p>
+{%endif%}
+
+{% if account_exists == False %}
+<h2 class="govuk-heading-m">This account has just been made:</h2>
+<p class="govuk-body">{{account_data|string}}</p>
+{% endif %}
```
Comment (gidsg) [inline app/discovery/templates/debug_continue.html:16]: Could this be an if / else?

File: app/discovery/templates/fund.html [hunk 4]
```diff
@@ -49,7 +49,8 @@ <h1 class="govuk-label-wrapper"> <label class="govuk-label govuk-label--m">
     </tbody>
   </table>
 
-  <a href={{round.application_url}} role="button" draggable="false" class="govuk-button govuk-button--start" data-module="govuk-button"> Start application <svg class="govuk-button__start-icon" xmlns="http://www.w3.org/2000/svg" width="17.5" height="19" viewBox="0 0 33 40" aria-hidden="true" focusable="false"> <path fill="currentColor" d="M0 0h13l20 20-20 20H0l20-20z"/> </svg></a>
+  <a href={{url_for("discovery_bp.email_route") + "?application_url=" + round.application_url|urlencode }} role="button" draggable="false" class="govuk-button govuk-button--start" data-module="govuk-button"> Start application <svg class="govuk-button__start-icon" xmlns="http://www.w3.org/2000/svg" width="17.5" height="19" viewBox="0 0 33 40" aria-hidden="true" focusable="false"> <path fill="currentColor" d="M0 0h13l20 20-20 20H0l20-20z"/> </svg></a>
```
Comment (harryyo) [inline app/discovery/templates/fund.html:52]: These args could be included in the url_for

File: app/discovery/models/data.py [hunk 5]
```diff
@@ -193,3 +195,77 @@ def list_data(json_data, data_func):
...
+    raw_params = {"email_address": email_address, "account_id": account_id}
+
+    params = {k: v for k, v in raw_params.items() if v is not None}
+
+    req.prepare_url(url, params)
+
+    response = requests.get(req.url)
+
+
+    if response.status_code == 204:
```
[the remark this answers is not in the document]
Comment (SentientHousePlant) [reply app/discovery/models/data.py]: Great catch thanks

File: app/discovery/routes.py [hunk 6]
```diff
@@ -75,4 +79,64 @@ def fund_rounds(fund_id):
...
+    if status_code == 200:
+
+        account_exists = True
+
+    if status_code == 204:
+
+        _, response_data = post_account(email)
+
+    return render_template(
+        "debug_continue.html",
```
[the remark this answers is not in the document]
Comment (SentientHousePlant) [reply app/discovery/routes.py:131]: Yes, it is soon to be replaced by @ramsharma-prog

File: app/discovery/templates/fund.html (hunk 4 above)
Comment (harryyo) [reply app/discovery/templates/fund.html:52]: Yep, any args after the route name which aren't specified as path variables on the route itself, get passed as query params conveniently.
https://stackoverflow.com/questions/23143827/redirecting-with-url-for-to-a-path-with-query-params-in-flask

Comment (SentientHousePlant) [reply app/discovery/templates/fund.html:52]: that is convenient thank you

File: tests/test_data_response.py [hunk 7]
```diff
@@ -69,3 +75,28 @@ def test_rounds_exist(flask_test_client):
...
+
+    random_string = "".join(
+        random.choices(string.ascii_uppercase + string.digits, k=10)
+    )
+    created_email = f"{random_string}@delete_me.com"
+    params = {"application_url": "www.google.com", "email": created_email}
+    req = PreparedRequest()
+    url = request.root_url + url_for("discovery_bp.account_info_route")
+    req.prepare_url(url, params)
+    _ = requests.get(req.url)
```
Comment (gidsg) [inline tests/test_data_response.py:98]: seems like this `_` assignment isn't doing anything, should we just remove it?

**Review by gidsg — APPROVED** [review]
LGTM 👍  as discussed it would be good to have a consistent approach to config but we can pick that up separately.

**Pull request merged**